Bgrenon partners guide (#1161)
neondatabase/website: content/docs/guides/partner-billing.md

---
title: Manage billing with consumption limits
subtitle: Learn how to set usage quotas per project with the Neon API
enableTableOfContents: true
isDraft: false
updatedOn: '2023-11-03T15:15:02.548Z'
---

When setting up your billing solution with Neon, you may want to impose some hard limits on how much storage or compute size a given project can consume. For example, you may want to cap how much usage your free tier users can consume versus pro or enterprise users. With the Neon API, you can use the `quota` key to set usage limits for a variety of consumption metrics. These limits act as thresholds after which all active computes for a project are [suspended](#what-happens-when-the-quota-is-met). 

## Metrics and quotas
By default, Neon tracks a variety of consumption metrics at the project level. If you want to set quotas (max limits) for these metrics, you need to explicitly [configure](#configuring-quotas) them. 

### Available metrics

Here are the relevant metrics that you can set project-level quotas for:

* `active_time_seconds`
* `compute_time_seconds`
* `written_data_bytes`
* `data_transfer_bytes` 

> These consumption metrics represent total cumulated usage across all branches and computes in a given project, accrued so far in a given monthly billing period. They are refreshed on a set day every month, on whichever date your new billing period starts. 

Reviewer: Is it any date or is it the first day of the month that a new billing period starts? We'll have to check. I always thought it was the first of the month, but I could be wrong.

Author: Revised to: 

"Metrics are refreshed on the first day of the following month."

To find the current usage level for any of these metrics, see [retrieving details about a project](#retrieving-details-about-a-project). You can read more about these metrics and how they impact billing [here](/docs/billing).

### Corresponding quotas

You can set quotas for these consumption metrics per project using the following Neon API URL:

```bash
/projects/{project_id}/settings/quota
```

The `quota` object includes the array of parameters used to set threshold limits. Their names generally match their corresponding metric:

> * `active_time_seconds` &#8212; Sets the maximum amount of wall-clock time allowed in total across all of a project's compute endpoints. This means the total elapsed time, in seconds, from start to finish for each transaction handled by the project's endpoints, accumulated before the one-month refresh date.

Reviewer: Does active time include the "auto-suspend delay" period as well?

> * `active_time_seconds` &#8212; Sets the maximum amount of wall-clock time allowed in total across all of a project's compute endpoints. This means the total elapsed time, in seconds, from start to finish for each transaction handled by the project's endpoints, accumulated before the one-month refresh date.

Reviewer: Active time might include the auto-suspend delay period as well.

Author: Sounds like no suspended time is excluded. From https://neon.tech/docs/introduction/billing#active-time

"...excludes time when computes are in an Idle state due to [auto-suspension](https://neon.tech/docs/reference/glossary#auto-suspend-compute) (scale-to-zero)."

> * `compute_time_seconds` &#8212; Sets the maximum amount of CPU seconds allowed in total across all of a project's compute endpoints. This differs from `active_time_seconds` in that it only counts the time the CPU spends executing a process. It excludes time spent waiting for external resources, I/O operations, or time spent in a blocked or idle state.

Reviewer: I am not sure about "This differs from `active_time_seconds` in that it only counts the time the CPU spends executing a process. It excludes time spent waiting for external resources, I/O operations, or time spent in a blocked or idle state."

The way I understand it is this, but maybe I am wrong:

If a compute with 4 vCPUs runs for 1 second, active_time_seconds=1.
If a compute with 4 vCPUs runs for 1 second, compute_time_seconds=4.

So, the larger your computes, the faster you'll burn through compute_time_seconds.

Author: I've revised it to:

Sets the maximum amount of CPU seconds allowed in total across all of a project's compute endpoints, including any endpoints deleted during the current billing period. Note that the larger your compute endpoints, the faster you'll consume through `compute_time_seconds`. For example, a compute instance with 0.25 vCPU uses 0.25 seconds of compute time for every second it's active. If it has 4 vCPUs, it uses 4 seconds of compute time in one active second.
   | vCPUs  | `active_time_seconds` | `compute_time_seconds` |
   |--------|-----------------------|------------------------|
   | 0.25   | 1                     | 0.25                   |
   | 4      | 1                     | 4                      |

* `data_transfer_bytes` &#8212; Sets the maximum amount of egress data, measured in bytes, that can be transferred out of Neon from across all of a project's branches using the proxy.

There is one additional `quota` parameter, `logical_size_bytes`, which applies to individual branches, not to the overall project. You can use `logical_size_bytes` to set the maximum size (measured in bytes) that any one individual branch is allowed to reach. Once this threshold is met, the compute for that particular branch (and _only_ that particular branch) is suspended. Note that this limit is _not_ refreshed once per month: it is a strict size limit that applies for the life of the branch.

### Sample quotas

Let's say you want to set limits for an application with two tiers, Trial and Pro, you might set limits like the following:

| Parameter (project)                 | Trial                     | Pro                      |
|------------------------|---------------------------|--------------------------|
| `active_time_seconds`  | 86,400 (1 day)            | 2,592,000 (30 days)      |

> | `compute_time_seconds` | 3,600 (1 hour)            | 50,000 (approx. 14 hours)|

Reviewer: If my assumptions about compute_time_seconds, I wonder if "1 hour" and "14 hours" makes sense. In our Free Tier, we allow a 1/4 vCPU to run 24/4, for example. So, the setting would be 1/4 of the active_time_seconds value.

Author: Does this look a bit better?

| Parameter (project)       | Trial  (.25 vCPU)               | Pro  (max 4 vCPU)                |
|---------------------------|---------------------------------|----------------------------------|
| `active_time_seconds`     | 633,600 (business month 22 days)| 2,592,000 (30 days)              |
| `compute_time_seconds`    | 158,400 (approx 44 hours)       | 10,368,000 (up to 4 times the active time)|
| `written_data_bytes`      | 1,000,000,000 (approx. 1 GiB)   | 50,000,000,000 (approx. 50 GiB)  |
| `data_transfer_bytes`     | 500,000,000 (approx. 500 MiB)   | 10,000,000,000 (approx. 10 GiB)  |

| Parameter (branch)        | Trial                           | Pro                              |
|---------------------------|---------------------------------|----------------------------------|
| `logical_size_bytes`      | 100,000,000 (approx. 100 MiB)   | 10,000,000,000 (approx. 10 GiB)  |

| `data_transfer_bytes`  | 500,000,000 (approx. 500 MiB)| 10,000,000,000 (approx. 10 GiB)  |

| Parameter (branch)             | Trial                    | Pro                      |
|------------------------|---------------------------|--------------------------|
| `logical_size_bytes`   | 100,000,000 (approx. 100 MiB)| 10,000,000,000 (approx. 10 GiB)  |

### Guidelines

Generally, the most effective quotas for controlling spend per project are those controlling maximum compute (`active_time_seconds` and `compute_time_seconds`) and maximum written storage (`written_data_bytes`). In practice, it is possible that `data_transfer_bytes` could introduce unintended logical constraints against your usage. For example, let's say you want to run a cleanup operation to reduce your storage. If part of this cleanup operation involves moving data across the network (for instance, to create an offsite backup before deletion), the `data_transfer_bytes` limit could prevent you from completing the operation &#8212; an undesirable situation where two measures meant to control cost interfere with one another.

## Suspending active computes

_**What happens when the quota is met?**_

> When any configured metric reaches its quota limit, all active computes for that project are automatically suspended. It is important to understand, this suspension is persistent. It works differently than the inactivity-based [autosuspend](/docs/guides/auto-suspend-guide), where computes restart at the next interaction: this suspend will _not_ restart at the next API call or incoming proxy connection. Without intervention, the suspension remains in place until the next billing period starts (`quota_reset_at`).

Reviewer: Should we mention who would have to intervene? Is it Neon Support or the partner?

Author: It's on the partner. Revised to:

"If you don't take explicit action otherwise, the suspension remains in place until the end of the current billing period starts (`consumption_period_end`)."

      "owner_id": "cf80d675-bb06-4e0c-9327-48ccfe84e7be"
    },
    {
      "id": "divine-tree-77657175",
      "platform_id": "aws",
      "region_id": "aws-us-east-2",
      "name": "RnameII",
      "provisioner": "k8s-pod",
      "default_endpoint_settings": {
        "autoscaling_limit_min_cu": 1,
        "autoscaling_limit_max_cu": 1,
        "suspend_timeout_seconds": 0
      },
      "settings": {
        "quota": {
          "active_time_seconds": 36000,
          "compute_time_seconds": 18000
        }
      },
      "pg_version": 15,
      "proxy_host": "us-east-2.aws.neon.tech",
      "branch_logical_size_limit": 204800,
      "branch_logical_size_limit_bytes": 214748364800,
      "store_passwords": true,
      "active_time": 0,
      "cpu_used_sec": 0,
      "creation_source": "console",
      "created_at": "2023-10-29T16:33:39Z",
      "updated_at": "2023-10-31T17:08:13Z",
      "synthetic_storage_size": 32671528,
      "quota_reset_at": "2023-11-01T00:00:00Z",
      "owner_id": "cf80d675-bb06-4e0c-9327-48ccfe84e7be"
    }
  ],
  "pagination": {
    "cursor": "patient-frost-50125040"
  }
}
```
</CodeBlock>
</details>

## Resetting a project after suspend

> Generally, projects remain suspended until the next billing period. It is good practice to notify your users when they are close to reaching a limit; if the user is then suspended and loses access to their database, it will not be unexpected. If you have configured no further actions, the user will have to wait until the next billing period starts to resume usage.

Reviewer: It says generally here. Are there exceptions?

Author: The exception -- I think -- is if the partner does something to reset. I'll remove "generally" though. It could be misleading.

---
title: Manage billing with consumption limits
subtitle: Learn how to set usage quotas per project with the Neon API
enableTableOfContents: true
isDraft: false
updatedOn: '2023-11-09T17:36:14.275Z'
---

> When setting up your billing solution with Neon, you may want to impose some hard limits on how much storage or compute size a given project can consume. For example, you may want to cap how much usage your free tier users can consume versus pro or enterprise users. With the Neon API, you can use the `quota` key to set usage limits for a variety of consumption metrics. These limits act as thresholds after which all active computes for a project are [suspended](#suspending-active-computes). 

Reviewer: ```suggestion
When setting up your billing solution with Neon, you may want to impose some hard limits on how much storage or compute resources a given project can consume. For example, you may want to cap how much usage your free tier users can consume versus pro or enterprise users. With the Neon API, you can use the `quota` key to set usage limits for a variety of consumption metrics. These limits act as thresholds after which all active computes for a project are [suspended](#suspending-active-computes). 
```

## Metrics and quotas
By default, Neon tracks a variety of consumption metrics at the project level. If you want to set quotas (max limits) for these metrics, you need to explicitly [configure](#configuring-quotas) them. 

### Available metrics

Here are the relevant metrics that you can set project-level quotas for:

* `active_time_seconds`
* `compute_time_seconds`
* `written_data_bytes`
* `data_transfer_bytes` 

These consumption metrics represent total cumulative usage across all branches and computes in a given project, accrued so far in a given monthly billing period. Metrics are refreshed on the first day of the following month, when the new billing period starts. 

Neon updates these metrics every 15 minutes but it could take up to 1 hour before they are reportable.

To find the current usage level for any of these metrics, see [retrieving details about a project](#retrieving-details-about-a-project). You can read more about these metrics and how they impact billing [here](/docs/billing).

### Corresponding quotas

You can set quotas for these consumption metrics per project using the following Neon API URL:

```bash

> /projects/{project_id}/settings/quota

Reviewer: I am wondering if providing this URL is useful.  Would it be more useful to say "using the quote object in the Create project or Update project API, and link to those methods?

Author: Revised to:

"You can set quotas for these consumption metrics per project using the `quote` object in the [Create project](https://api-docs.neon.tech/reference/createproject) or [Update project](https://api-docs.neon.tech/reference/updateproject) API."

```

The `quota` object includes an array of parameters used to set threshold limits. Their names generally match their corresponding metric:

> * `active_time_seconds` &#8212; Sets the maximum amount of wall-clock time allowed in total across all of a project's compute endpoints. This means the total elapsed time, in seconds, from start to finish for each transaction handled by the project's endpoints, accumulated so far during the current billing period.

Reviewer: I don't think this part is true, since it elapsed time can include the time that your compute is active because there is an autosuspend delay period:
"This means the total elapsed time, in seconds, from start to finish for each transaction handled by the project's endpoints, accumulated so far during the current billing period." 

I think it's just the total amount of time your project's compute endpoints are allowed to be active.

Author: Ah, right. The inactive period, before autosuspend kicks in, is counted. How about:

"Sets the maximum amount of time your project's total compute endpoints are allowed to be active during the current billing period. It excludes time when computes are in an `Idle` state due to [auto-suspension](/docs/reference/glossary#auto-suspend-compute)."

* `data_transfer_bytes` &#8212; Sets the maximum amount of egress data, measured in bytes, that can be transferred out of Neon from across all of a project's branches using the proxy.

There is one additional `quota` parameter, `logical_size_bytes`, which applies to individual branches, not to the overall project. You can use `logical_size_bytes` to set the maximum size (measured in bytes) that any one individual branch is allowed to reach. Once this threshold is met, the compute for that particular branch (and _only_ that particular branch) is suspended. Note that this limit is _not_ refreshed once per month: it is a strict size limit that applies for the life of the branch.

### Sample quotas

Let's say you want to set limits for an application with two tiers, Trial and Pro, you might set limits like the following:

| Parameter (project)       | Trial  (.25 vCPU)               | Pro  (max 4 vCPU)                |
|---------------------------|---------------------------------|----------------------------------|
| active_time_seconds     | 633,600 (business month 22 days)| 2,592,000 (30 days)              |

> | compute_time_seconds    | 158,400 (approx 44 hours)       | 10,368,000 (up to 4 times the active time)|

Reviewer: ```suggestion
| compute_time_seconds    | 158,400 (approx 44 hours)       | 10,368,000 (4 times the active time for 4 vCPUs)|
```

| Parameter (branch)        | Trial                           | Pro                              |
|---------------------------|---------------------------------|----------------------------------|
| logical_size_bytes      | 100,000,000 (approx. 100 MiB)   | 10,000,000,000 (approx. 10 GiB)  |

### Guidelines

Generally, the most effective quotas for controlling spend per project are those controlling maximum compute (`active_time_seconds` and `compute_time_seconds`) and maximum written storage (`written_data_bytes`). In practice, it is possible that `data_transfer_bytes` could introduce unintended logical constraints against your usage. For example, let's say you want to run a cleanup operation to reduce your storage. If part of this cleanup operation involves moving data across the network (for instance, to create an offsite backup before deletion), the `data_transfer_bytes` limit could prevent you from completing the operation &#8212; an undesirable situation where two measures meant to control cost interfere with one another.

## Suspending active computes

_**What happens when the quota is met?**_

> When any configured metric reaches its quota limit, all active computes for that project are automatically suspended. It is important to understand, this suspension is persistent. It works differently than the inactivity-based [autosuspend](/docs/guides/auto-suspend-guide), where computes restart at the next interaction: this suspend will _not_ restart at the next API call or incoming proxy connection. If you don't take explicit action otherwise, the suspension remains in place until the end of the current billing period starts (`consumption_period_end`).

Reviewer: ```suggestion
When any configured metric reaches its quota limit, all active computes for that project are automatically suspended. It is important to understand, this suspension is persistent. It works differently than the inactivity-based [autosuspend](/docs/guides/auto-suspend-guide), where computes restart at the next interaction: this suspend will _not_ restart at the next API call or incoming connection. If you don't take explicit action otherwise, the suspension remains in place until the end of the current billing period starts (`consumption_period_end`).
```

See [Querying metrics and quotas](#querying-metrics-and-quotas) to find your reset date, billing period, and other values related to the project's consumption.

<Admonition type="Note">

> Neon tracks these consumption metrics on a monthly cycle. If you want to track metrics on a specific time range, you need to store your metrics as snapshots. You can also use the Preview [Consumption API](#retrieving-metrics-for-all-projects) to collect metrics from across a range of billing periods. 

Reviewer: ```suggestion
Neon tracks these consumption metrics on a monthly cycle. If you want to track metrics on a different cycle, you need to take snapshots of your metrics at the desired interval and store the data externally. You can also use the Preview [Consumption API](#retrieving-metrics-for-all-projects) to collect metrics from across a range of billing periods. 
```

</Admonition>

## Configuring quotas

You can set quotas using the Neon API either in a `POST` when you create a project or a `PATCH` to update an existing project:
* [Set quotas when you create the project](#set-quotas-when-you-create-the-project)
* [Update an existing project](#update-an-existing-project)

### Set quotas when you create the project

> For performance reasons, you might want to configure these quotas at the same time that you create a new project for your user, reducing the number of API calls you need to make.

Reviewer: ```suggestion
For performance reasons, you might want to configure these quotas at the same time that you create a new project for your user using the [Create a project](https://api-docs.neon.tech/reference/createproject) API, reducing the number of API calls you need to make.
```

> Here is a sample `POST` in `curl` that creates a new project called `UserNew` and sets the `active_time_seconds` quota to a total allowed time of 10 hours (36,000 seconds) for the month, and a total allowed `compute_time_seconds` set to 2.5 hours (9,000 seconds) for the month.  

Reviewer: 2.5 hours makes sense if the compute is a .25 CU compute. Should we mention or does it complicate matters?

Author: I think it's good to include. How about this as an additional sentence?

"This 4:1 ratio between active and compute time is suitable for a fixed compute size of 0.25 vCPU."

<Admonition type="Warning" title="Preview API">
This functionality is part of the preview API and is subject to change in the future.
</Admonition>

Here is the URL in the Neon API where you can get details for all projects in your account:

```bash
https://console.neon.tech/api/v2/consumption/projects
```
This API endpoint accepts the following query parameters: `from`,`to`, `limit`, and `cursor`.

#### Set a date range across multiple billing periods

You can set `from` and `to` query parameters to define a time range that can span across multiple billing periods. 
* `from` — Sets the start date and time of the time period for which you are seeking metrics.
* `to` — Sets the end date and time for the  the interval for which you desire metrics.

The response is organized by project and billing period: one object per project, per active billing period within the range. For example, if you choose a 6-month time range you will get up to 6 objects for every project active within those months. The response includes any projects deleted within that time range.

If you do not include these parameters, the query defaults to the current consumption period. 

> Here is an example query that returns metrics from September 1st and December 1st, 2023.

Reviewer: ```suggestion
Here is an example query that returns metrics from September 1st and December 1st, 2023. Time values must be provided  in ISO 8601 format. You can use this [timestamp converter](https://www.timestamp-converter.com/).
```

> 

Reviewer: <CodeBlock shouldWrap>

> 

Reviewer: Use `<CodeBlock shouldWrap>` and `</CodeBlock>` to wrap this long example.

  "pagination": {
    "cursor": "divine-tree-77657175"
  }
}
```
</CodeBlock>
</details>

Key details:
* The `period_id` key and `previous_period_id` are unique values used to identify and connect periods across the time range. 
* The `period_start` and `period_end` keys show the dates for that particular billing period. A `null` value indicates that the object is for the current billing period.
* The `cursor` object under `pagination` shows the last project Id in the response. See more about pagination in the next section.

#### Control pagination for large result sets

To control pagination (number of results per response), you can include these query parameters:
* `limit` &#8212; sets the number of project objects to be included in the response
* `cursor` &#8212; by default, the response uses the project `id` from the last project in the list as the `cursor` value (included in the `pagination` object at the end of the response). Generally, it is up to the application to collect and use this cursor value when setting up the next request.

Here is an example `GET` request asking for the next 100 projects, starting with project id `divine-tree-77657175`:

> 

Reviewer: Use `<CodeBlock shouldWrap>` and `</CodeBlock>` to wrap this long example.

Reviewer: A question for Atli or someone who has worked with partners and "quotas": Should this document cover these project limits as well:

- autoscaling_limit_min_cu
- autoscaling_limit_max_cu
- suspend_timeout_seconds
- history_retention_seconds

---

When setting up your billing solution with Neon, you may want to impose some hard limits on how much storage or compute resources a given project can consume. For example, you may want to cap how much usage your free tier users can consume versus pro or enterprise users. With the Neon API, you can use the `quota` key to set usage limits for a variety of consumption metrics. These limits act as thresholds after which all active computes for a project are [suspended](#suspending-active-computes). 

## Metrics and quotas
By default, Neon tracks a variety of consumption metrics at the project level. If you want to set quotas (max limits) for these metrics, you need to explicitly [configure](#configuring-quotas) them. 

### Available metrics

Here are the relevant metrics that you can set project-level quotas for:

* `active_time_seconds`
* `compute_time_seconds`
* `written_data_bytes`
* `data_transfer_bytes` 

These consumption metrics represent total cumulative usage across all branches and computes in a given project, accrued so far in a given monthly billing period. Metrics are refreshed on the first day of the following month, when the new billing period starts. 

> There is an additional value you can set quota controls for: `logical_size`, which gives you the current size of a particular branch. 

Reviewer: So, Neon sets a branch size limit on Free and Pro (3GB and 200GB, respectively).

branch_logical_size_limit (MiB)
branch_logical_size_limit_bytes (Bytes)

Users can monitor their branch data size as it grows using the "Get branch" logical_size attribute.

Users can self-impose a branch size limit that applies to each branch individually using Create/Update project logical_size_bytes.

> There is an additional value you can set quota controls for: `logical_size`, which gives you the current size of a particular branch. 

Reviewer: And sorry for my confusion yesterday!

> There is an additional value you can set quota controls for: `logical_size`, which gives you the current size of a particular branch. 

Author: This all makes sense - and it's also nice seeing all these topically related values laid out in one flow like this. It's clarifying.
